pkg/permissions/permissions.go

```diff
@@ -0,0 +1,203 @@
+
+type checkerCtxKey struct{}
+
+// Permissions handles supporting authorization checks
+type Permissions struct {
+	enabled        bool
+	logger         *zap.SugaredLogger
+	client         *http.Client
+	url            *url.URL
+	skipper        middleware.Skipper
+	defaultChecker Checker
+}
+
+// Middleware produces echo middleware to handle authorization checks
+func (p *Permissions) Middleware() echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if !p.enabled || p.skipper(c) {
+				setCheckerContext(c, p.defaultChecker)
+
+				return next(c)
+			}
+
+			actor := echojwtx.Actor(c)
+			if actor == "" {
+				return echo.ErrUnauthorized.WithInternal(ErrNoAuthToken)
+			}
+
+			authHeader := strings.TrimSpace(c.Request().Header.Get(echo.HeaderAuthorization))
+
+			if len(authHeader) <= len(bearerPrefix) {
+				return echo.ErrUnauthorized.WithInternal(ErrInvalidAuthToken)
+			}
+
+			if !strings.EqualFold(authHeader[:len(bearerPrefix)], bearerPrefix) {
+				return echo.ErrUnauthorized.WithInternal(ErrInvalidAuthToken)
+			}
+
+			token := authHeader[len(bearerPrefix):]
+
+			setCheckerContext(c, p.checker(c, actor, token))
+
+			return next(c)
+		}
+	}
+}
+
+func (p *Permissions) checker(c echo.Context, actor, token string) Checker {
+	return func(ctx context.Context, resource gidx.PrefixedID, action string) error {
+		logger := p.logger.With("actor", actor, "resource", resource.String(), "action", action)
+
+		values := url.Values{}
+		values.Add("resource", resource.String())
+		values.Add("action", action)
+
+		url := *p.url
+		url.RawQuery = values.Encode()
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url.String(), nil)
+		if err != nil {
+			logger.Errorw("failed to create checker request", "error", err)
+
+			return errors.WithStack(err)
+		}
+
+		req.Header.Set(echo.HeaderAuthorization, c.Request().Header.Get(echo.HeaderAuthorization))
+
+		resp, err := p.client.Do(req)
+		if err != nil {
+			err = errors.WithStack(err)
+
+			logger.Errorw("failed to make request", "error", err)
+
+			return err
+		}
+
+		defer resp.Body.Close()
+
+		err = ensureValidServerResponse(resp)
+		if err != nil {
+			body, _ := io.ReadAll(resp.Body) //nolint:errcheck // ignore any errors reading as this is just for logging.
+
+			switch {
+			case errors.Is(err, ErrPermissionDenied):
+				logger.Warnw("unauthorized access to resource")
+			case errors.Is(err, ErrBadResponse):
+				logger.Errorw("bad response from server", "error", err, "response.status_code", resp.StatusCode, "response.body", string(body))
+			}
+
+			return err
+		}
+
+		logger.Debug("access granted to resource")
+
+		return nil
+	}
+}
+
+// New creates a new Permissions instance
+func New(config Config, options ...Option) (*Permissions, error) {
+	p := &Permissions{
+		enabled:        config.URL != "",
+		client:         http.DefaultClient,
```

Comment:
let's use a sane client with a really short timeout instead of the default client

Reply:
(unless you think its better for implementers to pass a client - but that feels like a 🦶 🔫

Reply:
yeah fair point